asm/ddot_test.go

```diff
@@ -0,0 +1,113 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package asm
+
+import (
+	"math"
+	"testing"
+)
+
+func TestDdotUnitary(t *testing.T) {
+	for i, test := range []struct {
+		xData []float64
+		yData []float64
+
+		want float64
+	}{
+		{
+			xData: []float64{2},
+			yData: []float64{-3},
+			want:  -6,
+		},
+		{
+			xData: []float64{0, 0, 1, 1, 2, -3, -4},
+			yData: []float64{0, 1, 0, 3, -4, 5, -6},
+			want:  4,
+		},
+	} {
+		x, _, _ := newGuardedVector(test.xData, 1)
+		y, _, _ := newGuardedVector(test.yData, 1)
+		got := DdotUnitary(x, y)
+
+		if math.IsNaN(got) {
+			t.Errorf("test %v: invalid memory access in DdotUnitary", i)
+			continue
+		}
+		if got != test.want {
```

Comment:
I thought about it but then neglected it. You are right that the low-level code here warrants extra precaution, so I will add it.